src/transformStateWithClones.js

```diff
@@ -7,7 +7,48 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const result = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        currentState = addProperties(currentState, action.extraData);
+
+        break;
+
+      case 'removeProperties':
+        currentState = removeProperties(currentState, action.keysToRemove);
+
+        break;
+
+      case 'clear':
+        currentState = clear(currentState);
+
+        break;
+    }
+    result.push(currentState);
```

Comment:
The current implementation pushes the `currentState` into the `result` array inside the loop, but it should be done outside of the `switch` block, after processing each action. This aligns with the rule: 'It is better to push a copy of the object to the array at the end of each loop cycle, but outside of the `switch` block, after the current action is processed.'